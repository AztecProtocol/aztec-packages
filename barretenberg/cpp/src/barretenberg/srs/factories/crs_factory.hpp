#pragma once
#include "barretenberg/common/mem.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include <cstddef>

namespace bb::pairing {
struct miller_lines;
} // namespace bb::pairing

namespace bb::srs::factories {

/**
 * A prover crs representation.
 */
template <typename Curve> class ProverCrs {
  public:
    virtual ~ProverCrs() = default;
    ;

    /**
     *  @brief Returns the monomial points in a form to be consumed by scalar_multiplication pippenger algorithm.
     */
    virtual typename Curve::AffineElement* get_monomial_points() = 0;
    virtual size_t get_monomial_size() const = 0;
};

template <typename Curve> class VerifierCrs {
  public:
    virtual ~VerifierCrs() = default;
};
template <> class VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    virtual Curve::G2AffineElement get_g2x() const = 0;
    /**
     * @brief As the G_2 element of the CRS is fixed, we can precompute the operations performed on it during the
     * pairing algorithm to optimize pairing computations.
     */
    virtual bb::pairing::miller_lines const* get_precomputed_g2_lines() const = 0;
    /**
     *  @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

template <> class VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    /**
     * @brief Returns the G_1 elements in the CRS after the pippenger point table has been applied on them
     *
     */
    virtual Curve::AffineElement* get_monomial_points() const = 0;
    virtual size_t get_monomial_size() const = 0;
    /**
     * @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

/**
 * A factory class to return the prover crs and verifier crs on request.
 * You can construct an empty placeholder factory, because composers need to be given a factory at construction time.
 */
template <typename Curve> class CrsFactory {
  public:
    CrsFactory() = default;
    CrsFactory(CrsFactory&& other) = default;
    virtual ~CrsFactory() = default;
    virtual std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t) { return nullptr; }
    virtual std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs([[maybe_unused]] size_t degree = 0)
    {
        return nullptr;
    }
};

} // namespace bb::srs::factories