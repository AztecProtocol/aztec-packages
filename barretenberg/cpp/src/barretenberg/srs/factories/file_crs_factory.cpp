#include "file_crs_factory.hpp"
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/pairing.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"

namespace bb::srs::factories {

FileVerifierCrs<curve::BN254>::FileVerifierCrs(std::string const& path, const size_t)
    : precomputed_g2_lines((bb::pairing::miller_lines*)(aligned_alloc(64, sizeof(bb::pairing::miller_lines) * 2)))
{
    using Curve = curve::BN254;
    auto point_buf = scalar_multiplication::point_table_alloc<Curve::AffineElement>(1);
    srs::IO<Curve>::read_transcript_g1(point_buf.get(), 1, path);
    srs::IO<curve::BN254>::read_transcript_g2(g2_x, path);
    bb::pairing::precompute_miller_lines(bb::g2::one, precomputed_g2_lines[0]);
    bb::pairing::precompute_miller_lines(g2_x, precomputed_g2_lines[1]);
    first_g1 = point_buf[0];
}

FileVerifierCrs<curve::BN254>::~FileVerifierCrs()
{
    aligned_free(precomputed_g2_lines);
}

FileVerifierCrs<curve::Grumpkin>::FileVerifierCrs(std::string const& path, const size_t num_points)
    : num_points(num_points)
{
    using Curve = curve::Grumpkin;
    monomials_ = scalar_multiplication::point_table_alloc<Curve::AffineElement>(num_points);
    srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
    scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
    first_g1 = monomials_[0];
};

curve::Grumpkin::AffineElement* FileVerifierCrs<curve::Grumpkin>::get_monomial_points() const
{
    return monomials_.get();
}

size_t FileVerifierCrs<curve::Grumpkin>::get_monomial_size() const
{
    return num_points;
}

template <typename Curve>
FileCrsFactory<Curve>::FileCrsFactory(std::string path, size_t initial_degree)
    : path_(std::move(path))
    , degree_(initial_degree)
{}

template <typename Curve>
std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> FileCrsFactory<Curve>::get_prover_crs(size_t degree)
{
    if (degree != degree_ || !prover_crs_) {
        prover_crs_ = std::make_shared<FileProverCrs<Curve>>(degree, path_);
        degree_ = degree;
    }
    return prover_crs_;
}

template <typename Curve>
std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> FileCrsFactory<Curve>::get_verifier_crs(size_t degree)
{
    if (degree != degree_ || !verifier_crs_) {
        verifier_crs_ = std::make_shared<FileVerifierCrs<Curve>>(path_, degree);
        degree_ = degree;
    }
    return verifier_crs_;
}

template class FileProverCrs<curve::BN254>;
template class FileProverCrs<curve::Grumpkin>;
template class FileCrsFactory<curve::BN254>;
template class FileCrsFactory<curve::Grumpkin>;

} // namespace bb::srs::factories
