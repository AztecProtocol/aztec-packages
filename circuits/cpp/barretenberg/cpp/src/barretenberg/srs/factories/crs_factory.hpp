#pragma once
#include "barretenberg/common/mem.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include <cstddef>

namespace barretenberg::pairing {
struct miller_lines;
} // namespace barretenberg::pairing

namespace barretenberg::srs::factories {

/**
 * A prover crs representation.
 */
template <typename Curve> class ProverCrs {
  public:
    virtual ~ProverCrs() = default;
    ;

    /**
     * Returns the monomial points in a form to be consumed by scalar_multiplication pippenger algorithm.
     */
    virtual typename Curve::AffineElement* get_monomial_points() = 0;
    virtual size_t get_monomial_size() const = 0;
};

class VerifierCrs {
  public:
    virtual ~VerifierCrs() = default;
    ;

    virtual barretenberg::g2::affine_element get_g2x() const = 0;

    virtual barretenberg::pairing::miller_lines const* get_precomputed_g2_lines() const = 0;
};

/**
 * A factory class to return the prover crs and verifier crs on request.
 * You can construct an empty placeholder factory, because composers need to be given a factory at construction time.
 */
class CrsFactory {
  public:
    CrsFactory() = default;
    CrsFactory(CrsFactory&& other) = default;
    virtual ~CrsFactory() = default;
    virtual std::shared_ptr<barretenberg::srs::factories::ProverCrs<curve::BN254>> get_prover_crs(size_t)
    {
        return nullptr;
    }
    virtual std::shared_ptr<barretenberg::srs::factories::VerifierCrs> get_verifier_crs() { return nullptr; }
};

} // namespace barretenberg::srs::factories
