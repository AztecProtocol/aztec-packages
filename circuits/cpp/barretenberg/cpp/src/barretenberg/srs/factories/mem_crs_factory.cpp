#include "mem_crs_factory.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/pairing.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"

namespace {

using namespace barretenberg;
using namespace barretenberg::srs::factories;

class MemProverCrs : public ProverCrs<curve::BN254> {
  public:
    MemProverCrs(std::vector<g1::affine_element> const& points)
        : num_points(points.size())
    {
        monomials_ = scalar_multiplication::point_table_alloc<g1::affine_element>(num_points);
        std::copy(points.begin(), points.end(), monomials_.get());
        scalar_multiplication::generate_pippenger_point_table<curve::BN254>(
            monomials_.get(), monomials_.get(), num_points);
    }

    g1::affine_element* get_monomial_points() override { return monomials_.get(); }

    size_t get_monomial_size() const override { return num_points; }

  private:
    size_t num_points;
    std::shared_ptr<g1::affine_element[]> monomials_;
};

class MemVerifierCrs : public VerifierCrs {
  public:
    MemVerifierCrs(g2::affine_element const& g2_point)
        : g2_x(g2_point)
    {
        precomputed_g2_lines =
            (pairing::miller_lines*)(aligned_alloc(64, sizeof(barretenberg::pairing::miller_lines) * 2));
        barretenberg::pairing::precompute_miller_lines(barretenberg::g2::one, precomputed_g2_lines[0]);
        barretenberg::pairing::precompute_miller_lines(g2_x, precomputed_g2_lines[1]);
    }

    ~MemVerifierCrs() { aligned_free(precomputed_g2_lines); }

    g2::affine_element get_g2x() const override { return g2_x; }

    pairing::miller_lines const* get_precomputed_g2_lines() const override { return precomputed_g2_lines; }

  private:
    g2::affine_element g2_x;
    pairing::miller_lines* precomputed_g2_lines;
};

} // namespace

namespace barretenberg::srs::factories {

MemCrsFactory::MemCrsFactory(std::vector<g1::affine_element> const& points, g2::affine_element const g2_point)
    : prover_crs_(std::make_shared<MemProverCrs>(points))
    , verifier_crs_(std::make_shared<MemVerifierCrs>(g2_point))
{}

std::shared_ptr<barretenberg::srs::factories::ProverCrs<curve::BN254>> MemCrsFactory::get_prover_crs(size_t)
{
    return prover_crs_;
}

std::shared_ptr<barretenberg::srs::factories::VerifierCrs> MemCrsFactory::get_verifier_crs()
{
    return verifier_crs_;
}

} // namespace barretenberg::srs::factories