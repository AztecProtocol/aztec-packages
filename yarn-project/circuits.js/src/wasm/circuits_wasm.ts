import { numToUInt32BE } from '@aztec/foundation/serialize';
import { fileURLToPath } from '@aztec/foundation/url';
import { IWasmModule, WasmModule } from '@aztec/foundation/wasm';

import isNode from 'detect-node';
import { readFile } from 'fs/promises';
import { dirname, join } from 'path';

import { Crs } from '../crs/index.js';

const NAME = '/aztec3-circuits';
const CODE_PATH = isNode
  ? join(dirname(fileURLToPath(import.meta.url)), `../../resources/${NAME}.wasm`)
  : `${NAME}.wasm`;
const MAX_SRS_POINTS = 50000;

/**
 * Get the text of a binary file, either locally or on the web.
 * @param path - Path to the WASM binary.
 * @returns The binary buffer.
 */
async function fetchBinary(path: string) {
  if (isNode) {
    return await readFile(path);
  } else {
    const res = await fetch(path);
    return Buffer.from(await res.arrayBuffer());
  }
}
const BB_JS_NYI_ERROR =
  'NOT YET IMPLEMENTED - needed for proofs, plan is to use barretenberg.js from NPM for proofs. See https://github.com/AztecProtocol/aztec-packages/issues/781';
/**
 * A low-level wrapper for an instance of Aztec3 Circuits WASM.
 */
export class CircuitsWasm implements IWasmModule {
  static instance: Promise<CircuitsWasm>;

  /**
   * Creates a Circuits module from a properly configured WasmModule.
   * Not meant for public use.
   *
   * @param wasm - The module configured in CircuitsWasm.new().
   */
  private constructor(private wasm: WasmModule) {}

  /**
   * Get a singleton instance of the module.
   * @returns The singleton.
   */
  public static get(): Promise<CircuitsWasm> {
    if (!this.instance) this.instance = CircuitsWasm.new();
    return this.instance;
  }

  /**
   * Create and initialize a Circuits module. Not meant for public use.
   *
   * Has 30 pages by default. 30*2**16 \> 1mb stack size plus other overheads.
   * 8192 maximum by default. 512mb.
   * @param initial - Initial memory pages.
   * @param maximum - Max memory pages.
   * @param loggerName - The logger name.
   * @returns The wrapper.
   */
  private static async new(initial = 30, maximum = 8192, loggerName = 'aztec:wasm'): Promise<CircuitsWasm> {
    const wasm = new WasmModule(
      await fetchBinary(CODE_PATH),
      module => ({
        /**
         * Log a string from wasm.
         * @param addr - The string address to log.
         */
        logstr(addr: number) {
          const rawStr = wasm.getMemoryAsString(addr);
          const m = wasm.getMemory();
          const str = `${rawStr} (mem: ${(m.length / (1024 * 1024)).toFixed(2)}MB)`;
          if (str.startsWith('abort: ') || str.startsWith('important: ')) {
            // we explicitly want to route to console.log for every abort message to not miss them:
            // eslint-disable-next-line no-console
            console.log(str);
          }
          wasm.getLogger()(str);
        },
        memory: module.getRawMemory(),
        // eslint-disable-next-line camelcase
        set_data: () => {
          wasm.getLogger()('set_data: NYI');
        },
        // eslint-disable-next-line camelcase
        get_data: () => {
          throw new Error(BB_JS_NYI_ERROR);
        },
      }),
      loggerName,
    );
    await wasm.init(initial, maximum);
    await CircuitsWasm.initializeSrs(wasm);
    return new CircuitsWasm(wasm);
  }

  /**
   * Ensures the global SRS is initialized.
   * Currently used in VK serialization and will be used in proofs.
   * TODO(AD): proof should use external bb.js
   * TODO(AD): revisit when SRS should be initialized
   * @param wasm - The WASM module.
   */
  private static async initializeSrs(wasm: WasmModule) {
    const crs = new Crs(MAX_SRS_POINTS);
    await crs.init();
    const g1Buf = wasm.call('bbmalloc', crs.getG1Data().length);
    wasm.writeMemory(g1Buf, crs.getG1Data());
    const g1SizeBuf = wasm.call('bbmalloc', 4);
    wasm.writeMemory(g1SizeBuf, numToUInt32BE(crs.numPoints));
    const g2Buf = wasm.call('bbmalloc', crs.getG2Data().length);
    wasm.writeMemory(g2Buf, crs.getG2Data());
    wasm.call('srs_init_srs', g1Buf, g1SizeBuf, g2Buf);
    wasm.call('bbfree', g1Buf);
    wasm.call('bbfree', g1SizeBuf);
    wasm.call('bbfree', g2Buf);
  }

  /**
   * Retrieve the exports object of the CircuitsWasm module.
   *
   * @returns An object containing exported functions and properties.
   */
  public exports(): any {
    return this.wasm.exports();
  }

  /**
   * Get a slice of memory between two addresses.
   * @param start - The start address.
   * @param end - The end address.
   * @returns A Uint8Array view of memory.
   */
  public getMemorySlice(start: number, end: number) {
    return this.wasm.getMemorySlice(start, end);
  }

  /**
   * Write data into the heap.
   * @param arr - The data to write.
   * @param offset - The address to write data at.
   */
  public writeMemory(offset: number, arr: Uint8Array) {
    this.wasm.writeMemory(offset, arr);
  }

  /**
   * Get memory as string.
   * @param offset - The address to get null-terminated string data from.
   * @returns JS string.
   */
  public getMemoryAsString(offset: number) {
    return this.wasm.getMemoryAsString(offset);
  }

  /**
   * Calls into the WebAssembly.
   * @param name - The method name.
   * @param args - The arguments to the method.
   * @returns The numeric integer or address result.
   */
  public call(name: string, ...args: any): number {
    return this.wasm.call(name, ...args);
  }
}
