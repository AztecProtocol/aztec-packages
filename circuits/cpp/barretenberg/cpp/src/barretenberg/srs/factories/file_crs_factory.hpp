#pragma once
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "crs_factory.hpp"
#include <cstddef>
#include <utility>

namespace barretenberg::srs::factories {

template <typename Curve> class FileProverCrs : public ProverCrs<Curve> {
  public:
    FileProverCrs(const size_t num_points, std::string const& path);

    typename Curve::AffineElement* get_monomial_points() { return monomials_.get(); }

    size_t get_monomial_size() const { return num_points; }

  private:
    size_t num_points;
    std::shared_ptr<typename Curve::AffineElement[]> monomials_;
};

class FileVerifierCrs : public VerifierCrs {
  public:
    FileVerifierCrs(std::string const& path);

    ~FileVerifierCrs();

    g2::affine_element get_g2x() const override;

    pairing::miller_lines const* get_precomputed_g2_lines() const override;

  private:
    g2::affine_element g2_x;
    pairing::miller_lines* precomputed_g2_lines;
};

/**
 * Create reference strings given a path to a directory of transcript files.
 */
class FileCrsFactory : public CrsFactory {
  public:
    FileCrsFactory(std::string path, size_t initial_degree = 0);
    FileCrsFactory(FileCrsFactory&& other) = default;

    std::shared_ptr<barretenberg::srs::factories::ProverCrs<curve::BN254>> get_prover_crs(size_t degree) override;

    std::shared_ptr<barretenberg::srs::factories::VerifierCrs> get_verifier_crs() override;

  private:
    std::string path_;
    size_t degree_;
    std::shared_ptr<barretenberg::srs::factories::ProverCrs<curve::BN254>> prover_crs_;
    std::shared_ptr<barretenberg::srs::factories::VerifierCrs> verifier_crs_;
};

extern template class FileProverCrs<curve::BN254>;
extern template class FileProverCrs<curve::Grumpkin>;

} // namespace barretenberg::srs::factories
