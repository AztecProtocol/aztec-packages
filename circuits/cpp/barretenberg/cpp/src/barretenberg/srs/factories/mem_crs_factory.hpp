#pragma once
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "crs_factory.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include <utility>
#include <cstddef>

namespace barretenberg::srs::factories {

/**
 * Create reference strings given pointers to in memory buffers.
 */
class MemCrsFactory : public CrsFactory {
  public:
    MemCrsFactory(std::vector<g1::affine_element> const& points, g2::affine_element const g2_point);
    MemCrsFactory(MemCrsFactory&& other) = default;

    std::shared_ptr<barretenberg::srs::factories::ProverCrs<curve::BN254>> get_prover_crs(size_t degree) override;

    std::shared_ptr<barretenberg::srs::factories::VerifierCrs> get_verifier_crs() override;

  private:
    std::shared_ptr<barretenberg::srs::factories::ProverCrs<curve::BN254>> prover_crs_;
    std::shared_ptr<barretenberg::srs::factories::VerifierCrs> verifier_crs_;
};

} // namespace barretenberg::srs::factories
