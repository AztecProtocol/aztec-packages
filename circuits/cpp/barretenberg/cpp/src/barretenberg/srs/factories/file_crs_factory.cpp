#include "file_crs_factory.hpp"
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/pairing.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"

namespace barretenberg::srs::factories {

template <typename Curve>
FileProverCrs<Curve>::FileProverCrs(const size_t num_points, std::string const& path)
    : num_points(num_points)
{
    monomials_ = scalar_multiplication::point_table_alloc<typename Curve::AffineElement>(num_points);

    srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
    scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
}

FileVerifierCrs::FileVerifierCrs(std::string const& path)
    : precomputed_g2_lines(
          (barretenberg::pairing::miller_lines*)(aligned_alloc(64, sizeof(barretenberg::pairing::miller_lines) * 2)))
{

    srs::IO<curve::BN254>::read_transcript_g2(g2_x, path);
    barretenberg::pairing::precompute_miller_lines(barretenberg::g2::one, precomputed_g2_lines[0]);
    barretenberg::pairing::precompute_miller_lines(g2_x, precomputed_g2_lines[1]);
}

FileVerifierCrs::~FileVerifierCrs()
{
    aligned_free(precomputed_g2_lines);
}

g2::affine_element FileVerifierCrs::get_g2x() const
{
    return g2_x;
}

pairing::miller_lines const* FileVerifierCrs::get_precomputed_g2_lines() const
{
    return precomputed_g2_lines;
}

FileCrsFactory::FileCrsFactory(std::string path, size_t initial_degree)
    : path_(std::move(path))
    , degree_(initial_degree)
    , verifier_crs_(std::make_shared<FileVerifierCrs>(path_))
{}

std::shared_ptr<barretenberg::srs::factories::ProverCrs<curve::BN254>> FileCrsFactory::get_prover_crs(size_t degree)
{
    if (degree != degree_ || !prover_crs_) {
        prover_crs_ = std::make_shared<FileProverCrs<curve::BN254>>(degree, path_);
        degree_ = degree;
    }
    return prover_crs_;
}

std::shared_ptr<barretenberg::srs::factories::VerifierCrs> FileCrsFactory::get_verifier_crs()
{
    return verifier_crs_;
}

template class FileProverCrs<curve::BN254>;
template class FileProverCrs<curve::Grumpkin>;

} // namespace barretenberg::srs::factories