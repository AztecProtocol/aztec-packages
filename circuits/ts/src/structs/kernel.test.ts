import { expectSerializeToMatchSnapshot } from '../tests/expectSerialize.js';
import { makePreviousKernelData } from '../tests/factories.js';
import { writeGlobalVerifierReferenceString } from '../tests/writeGlobalVerifierReferenceString.js';
import { CircuitsWasm } from '../wasm/circuits_wasm.js';

describe('structs/kernel', () => {
  it(`serializes and prints previous_kernel_data`, async () => {
    const wasm = await CircuitsWasm.new();
    const previousKernelData = makePreviousKernelData();
    await writeGlobalVerifierReferenceString(wasm, /* example circuit size */ 100);
    await expectSerializeToMatchSnapshot(
      previousKernelData.toBuffer(),
      'abis__test_roundtrip_serialize_previous_kernel_data',
      wasm,
    );
  });
});
