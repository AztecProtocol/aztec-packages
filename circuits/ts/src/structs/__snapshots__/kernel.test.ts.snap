// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`structs/kernel serializes and prints previous_kernel_data 1`] = `
"public_inputs: end:
aggregation_object:
P0: { 0x1, 0x2 }
P1: { 0x101, 0x102 }
public_inputs:
[
 0x3
 0x4
 0x5
 0x6
]
proof_witness_indices:
[ 7 8 9 10 11 12 ]
has_data: 0

private_call_count: 0xd
new_commitments:
[ 0x101 0x102 0x103 0x104 0x105 0x106 0x107 0x108 0x109 0x10a 0x10b 0x10c 0x10d 0x10e 0x10f 0x110 ]
new_nullifiers:
[ 0x201 0x202 0x203 0x204 0x205 0x206 0x207 0x208 0x209 0x20a 0x20b 0x20c 0x20d 0x20e 0x20f 0x210 ]
private_call_stack:
[ 0x301 0x302 0x303 0x304 0x305 0x306 0x307 0x308 ]
public_call_stack:
[ 0x401 0x402 0x403 0x404 0x405 0x406 0x407 0x408 ]
l1_msg_stack:
[ 0x501 0x502 0x503 0x504 ]
new_contracts:
[ contract_address: 0x601
portal_contract_address: 0x202020202020202020202020202020202020202
function_tree_root: 0x603
 contract_address: 0x602
portal_contract_address: 0x303030303030303030303030303030303030303
function_tree_root: 0x604
 contract_address: 0x603
portal_contract_address: 0x404040404040404040404040404040404040404
function_tree_root: 0x605
 contract_address: 0x604
portal_contract_address: 0x505050505050505050505050505050505050505
function_tree_root: 0x606
 contract_address: 0x605
portal_contract_address: 0x606060606060606060606060606060606060606
function_tree_root: 0x607
 contract_address: 0x606
portal_contract_address: 0x707070707070707070707070707070707070707
function_tree_root: 0x608
 contract_address: 0x607
portal_contract_address: 0x808080808080808080808080808080808080808
function_tree_root: 0x609
 contract_address: 0x608
portal_contract_address: 0x909090909090909090909090909090909090909
function_tree_root: 0x60a
 ]
optionally_revealed_data:
[ call_stack_item_hash: 0x701
function_data:
function_selector: 702
is_private: 1
is_constructor: 1

emitted_events:
[ 0x801 0x802 0x803 0x804 ]
vk_hash: 0x703
portal_contract_address: 0x404040404040404040404040404040404040404
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x702
function_data:
function_selector: 703
is_private: 1
is_constructor: 1

emitted_events:
[ 0x802 0x803 0x804 0x805 ]
vk_hash: 0x704
portal_contract_address: 0x505050505050505050505050505050505050505
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x703
function_data:
function_selector: 704
is_private: 1
is_constructor: 1

emitted_events:
[ 0x803 0x804 0x805 0x806 ]
vk_hash: 0x705
portal_contract_address: 0x606060606060606060606060606060606060606
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 call_stack_item_hash: 0x704
function_data:
function_selector: 705
is_private: 1
is_constructor: 1

emitted_events:
[ 0x804 0x805 0x806 0x807 ]
vk_hash: 0x706
portal_contract_address: 0x707070707070707070707070707070707070707
pay_fee_from_l1: 1
pay_fee_from_public_l2: 0
called_from_l1: 1
called_from_public_l2: 0
 ]

constants:
old_tree_roots: private_data_tree_root: 0x101
nullifier_tree_root: 0x102
contract_tree_root: 0x103
private_kernel_vk_tree_root: 0x104

tx_context: is_fee_payment_tx: 0
is_rebate_payment_tx: 0
is_contract_deployment_tx: 1
contract_deployment_data: 
constructor_vk_hash: 0x105
function_tree_root: 0x106
contract_address_salt: 0x107
portal_contract_address: 0x808080808080808080808080808080808080808


is_private: 1

proof: [ 81 81 81 81 81 81 81 81 81 81 81 81 81 81 81 81 ]
vk: composer_type: 0
circuit_size: 101
num_public_inputs: 102
commitments: [
 A: { 0x200, 0x300 }
]
contains_recursive_proof: 1
recursive_proof_public_input_indices: [ 400 401 402 403 404 ]

vk_index: 66
vk_path: [ 0x1000 0x1001 0x1002 ]
"
`;
